airflow/models/trigger.py

```diff
@@ -115,7 +115,7 @@ def submit_event(cls, trigger_id, event, session=None):
             TaskInstance.trigger_id == trigger_id, TaskInstance.state == State.DEFERRED
         ):
             # Add the event's payload into the kwargs for the task
-            next_kwargs = task_instance.next_kwargs or {}
+            next_kwargs = dict(task_instance.next_kwargs) if task_instance.next_kwargs else {}
```

Comment:
@kaxil  sqlalchemy doesn't update next_kwargs column if same dict object is modified,
I've updated test_trigger.py's test_submit_event to capture this

Reply:
Seems that is indeed unfortunate way how some objects are treated by SQLAlchemy. Basically changes to object are not seen as "changes" unles we explicitly replace the object or mark it as changed.

But likely there is a better way of handling it:

https://docs.sqlalchemy.org/en/14/orm/extensions/mutable.html

Seems that we could set the object as Mutable and it would solve the problem.

Reply:
(and it prevent similar problem from happening in other places).

Reply:
:( SqlAlchemy

Reply:
This is only a problem when you mutate a structure (dict, list etc) in place and is sort of a limitation of Python -- nothing is called on the object to know it's changed (without doing _very_ expensive checks all the time)

Reply:
`ExtendedJSON` is also used in `TaskMap.keys` and `DbCallbackRequest.callback_data` so this might suffer from same issue if they mutate the dicts too.

https://github.com/apache/airflow/blob/152b646564c8eff1239f29e91aec5e0f89b28353/airflow/models/db_callback_request.py#L37
https://github.com/apache/airflow/blob/152b646564c8eff1239f29e91aec5e0f89b28353/airflow/models/taskmap.py#L61

cc @uranusjr @ashb

Reply:
https://github.com/edelooff/sqlalchemy-json

Reply:
(but in this case doing what we have in this PR may be the better option)

Reply:
We can apply this change too

```diff
diff --git a/airflow/models/taskinstance.py b/airflow/models/taskinstance.py
index 5b53916ac..fc4003b4f 100644
--- a/airflow/models/taskinstance.py
+++ b/airflow/models/taskinstance.py
@@ -69,6 +69,7 @@ from sqlalchemy import (
     tuple_,
 )
 from sqlalchemy.ext.associationproxy import association_proxy
+from sqlalchemy.ext.mutable import MutableDict
 from sqlalchemy.orm import reconstructor, relationship
 from sqlalchemy.orm.attributes import NO_VALUE, set_committed_value
 from sqlalchemy.orm.exc import NoResultFound
@@ -453,7 +454,7 @@ class TaskInstance(Base, LoggingMixin):
     # The method to call next, and any extra arguments to pass to it.
     # Usually used when resuming from DEFERRED.
     next_method = Column(String(1000))
-    next_kwargs = Column(ExtendedJSON)
+    next_kwargs = Column(MutableDict.as_mutable(ExtendedJSON))

     # If adding new fields here then remember to add them to
     # refresh_from_db() or they won't display in the UI correctly
diff --git a/tests/models/test_trigger.py b/tests/models/test_trigger.py
index 99cd71f76..ed6f635a9 100644
--- a/tests/models/test_trigger.py
+++ b/tests/models/test_trigger.py
@@ -101,6 +101,9 @@ def test_submit_event(session, create_task_instance):
     session.commit()
     # Call submit_event
     Trigger.submit_event(trigger.id, TriggerEvent(42), session=session)
+    # commit changes made by submit event and expire all cache to read from db.
+    session.commit()
+    session.expire_all()
     # Check that the task instance is now scheduled
     updated_task_instance = session.query(TaskInstance).one()
     assert updated_task_instance.state == State.SCHEDULED
```

Reply:
Other uses are fine -- this is the only one where we mutate a row in place. Others are create only/set new dict.

Reply:
in that, I propose using the diff I have above to be explicit instead of using `dict` as it is in the current PR

Reply:
Hmm is this the same cause to https://github.com/apache/airflow/issues/22245?